x/programs/rust/examples/token/src/lib.rs

```diff
@@ -16,108 +13,179 @@ pub enum StateKeys {
     Symbol,
     /// The balance of the token by address. Key prefix 0x3 + address.
     Balance(Address),
+    /// The allowance of the token by owner and spender. Key prefix 0x4 + owner + spender.
+    Allowance(Address, Address),
+    // Original owner of the token
+    Owner,
+}
+
+// Returns the owner of the token
+pub fn get_owner(program: &Program<StateKeys>) -> Address {
+    program
+        .state()
+        .get::<Address>(StateKeys::Owner)
+        .expect("failure")
+        .unwrap()
+}
+
+// Checks if the caller is the owner of the token
+// If the caller is not the owner, the program will panic
+pub fn owner_check(program: &Program<StateKeys>, actor: Address) {
+    assert!(
+        get_owner(program) == actor,
+        "caller is required to be owner"
+    )
 }
 
 /// Initializes the program with a name, symbol, and total supply.
 #[public]
-pub fn init(context: Context<StateKeys>) {
+pub fn init(context: Context<StateKeys>, name: String, symbol: String) {
     let program = context.program();
+    let actor = context.actor();
 
-    // set total supply
-    program
-        .state()
-        .store_by_key(StateKeys::TotalSupply, &INITIAL_SUPPLY)
-        .expect("failed to store total supply");
-
-    // set token name
     program
         .state()
-        .store_by_key(StateKeys::Name, b"WasmCoin")
-        .expect("failed to store coin name");
-
-    // set token symbol
+        .store_by_key(StateKeys::Owner, &actor)
```

Comment:
the values in each pair passed into `store` must be the same type. Can change this but it may conflict with the `state` stuff you are working on and might not be relevant soon

Reply:
FFS... I think that was a bit of an oversight on my part. Yeah, no need to make changes until after my stuff lands